sdk/resourcemanager/Azure.ResourceManager/docs/CONTRIBUTING.md

```diff
@@ -34,23 +34,79 @@ We use [NUnit 3][nunit] as our testing framework.
 
 [Azure.Core.TestFramework's testing framework][core_tests] provides a set of reusable primitives that simplify writing tests for new Azure SDK libraries.
 
-### Sync/Async testing
+### Testing project structure
+
+With the help of [Azure.ResourceManager.Template](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/templates/Azure.ResourceManager.Template), the basic testing project and file structure(see following for details) will be generated under `sdk\<service name>\<package name>\tests` directory.
+
+```text
+sdk\<service name>\<package name>\tests\Azure.ResourceManager.<service>.Tests.csproj
+sdk\<service name>\<package name>\tests\<service>ManagementTestBase.cs
+sdk\<service name>\<package name>\tests\<service>ManagementTestEnvironment.cs
+sdk\<service name>\<package name>\tests\SessionRecords
+sdk\<service name>\<package name>\tests\Scenario
+```
+
+### Writing scenario tests
+
+All the public APIs that are exposed to the customer need to be tested and they are distributed in the following three kinds of code files as shown below. Accordingly, you'd better put the tests of the same file's APIs in a separate file under the `Scenario` folder, thus facilitating the subsequent maintenance.
+
+```text
+sdk\<service name>\<package name>\src\Generated\Extensions\<service>Extensions.cs
+sdk\<service name>\<package name>\src\Generated\<resource>Collection.cs
+sdk\<service name>\<package name>\src\Generated\<resource>Resource.cs
+```
+
+For instance, if you want to test the `CreateOrUpdate` method in ExampleCollection.cs, the corresponding test file will be like:
+
+```csharp
+namespace Azure.ResourceManager.Service.Tests
+{
+    public class ExampleCollectionTests : ServiceManagementTestBase
+    {
+        public ExampleCollectionTests(bool isAsync)
+            : base(isAsync)//, RecordedTestMode.Record)
+        {
+        }
+
+        [TestCase]
+        [RecordedTest]
+        public async Task CreateOrUpdate()
+        {
+            SubscriptionResource subscription = await Client.GetDefaultSubscriptionAsync();
+            ResourceGroupResource rg = await CreateResourceGroup(subscription, "testRg", AzureLocation.WestUS);
+            string resourceName = Recording.GenerateAssetName("resource");
+            ExampleResource resource = await rg.GetExamples().CreateOrUpdateAsync(WaitUntil.Completed, resourceName, new ExampleData());
+            Assert.AreEqual(resourceName, resource.Data.Name);
+        }
+```
+
+We expose all of our APIs with both sync and async variants. To avoid writing each of our tests twice, we automatically rewrite our async API calls into their sync equivalents. Simply write your tests using only async APIs and the test framework will wrap the client with a proxy that forwards everything to the sync overloads automatically. Visual Studio's test runner will show `*TestClass(True)` for the async variants and `*TestClass(False)` for the sync variants.
```

Comment:
I think it's not `rewrite`, but `convert at runtime`.

Reply:
changed

Reply:
changed